modules/call-dispatch/src/lib.rs

```diff
@@ -60,75 +105,117 @@ pub trait Trait: frame_system::Trait {
 	}
 }
 
-impl<T: Trait> MessageDispatch<T::MessageId> for Module<T> {
-	type Message = (SpecVersion, Weight, <T as Trait>::Call);
+impl<T: Trait<I>, I: Instance> MessageDispatch<T::MessageId> for Module<T, I> {
+	type Message = MessagePayload<
+		T::SourceChainAccountPublic,
+		T::TargetChainAccountPublic,
+		T::TargetChainSignature,
+		<T as Trait<I>>::Call,
+	>;
 
 	fn dispatch(bridge: InstanceId, id: T::MessageId, message: Self::Message) -> Weight {
-		let (spec_version, weight, call) = message;
-
 		// verify spec version
 		// (we want it to be the same, because otherwise we may decode Call improperly)
 		let expected_version = <T as frame_system::Trait>::Version::get().spec_version;
-		if spec_version != expected_version {
+		if message.spec_version != expected_version {
 			frame_support::debug::trace!(
 				"Message {:?}/{:?}: spec_version mismatch. Expected {:?}, got {:?}",
 				bridge,
 				id,
 				expected_version,
-				spec_version,
+				message.spec_version,
 			);
-			Self::deposit_event(Event::<T>::MessageVersionSpecMismatch(
+			Self::deposit_event(RawEvent::MessageVersionSpecMismatch(
 				bridge,
 				id,
 				expected_version,
-				spec_version,
+				message.spec_version,
 			));
 			return DEPOSIT_EVENT_WEIGHT;
 		}
 
 		// verify weight
 		// (we want passed weight to be at least equal to pre-dispatch weight of the call
 		// because otherwise Calls may be dispatched at lower price)
-		let dispatch_info = call.get_dispatch_info();
+		let dispatch_info = message.call.get_dispatch_info();
 		let expected_weight = dispatch_info.weight;
-		if weight < expected_weight {
+		if message.weight < expected_weight {
 			frame_support::debug::trace!(
 				"Message {:?}/{:?}: passed weight is too low. Expected at least {:?}, got {:?}",
 				bridge,
 				id,
 				expected_weight,
-				weight,
+				message.weight,
 			);
-			Self::deposit_event(Event::<T>::MessageWeightMismatch(bridge, id, expected_weight, weight));
+			Self::deposit_event(RawEvent::MessageWeightMismatch(
+				bridge,
+				id,
+				expected_weight,
+				message.weight,
+			));
 			return DEPOSIT_EVENT_WEIGHT;
 		}
 
+		// prepare dispatch origin
+		let origin_account = match message.origin {
+			CallOrigin::BridgeAccount => bridge_account_id(bridge, CALL_DISPATCH_MODULE_PREFIX),
+			CallOrigin::RealAccount(source_public, target_public, target_signature) => {
+				let mut signed_message = Vec::new();
+				message.call.encode_to(&mut signed_message);
+				source_public.encode_to(&mut signed_message);
+
+				let target_account = target_public.into_account();
+				if !target_signature.verify(&signed_message[..], &target_account) {
+					frame_support::debug::trace!(
+						"Message {:?}/{:?}: origin proof is invalid. Expected account: {:?} from signature: {:?}",
+						bridge,
+						id,
+						target_account,
+						target_signature,
+					);
+					Self::deposit_event(RawEvent::MessageSignatureMismatch(bridge, id));
+					return DEPOSIT_EVENT_WEIGHT;
```

Comment:
Shouldn't we charge a bit more due to signature verification?

Reply:
This (returned `Weight`) is actually removed in #339